File: com.trollworks.gcs/src/com/trollworks/gcs/modifier/ModifierWeightValueType.java
/*
 * Copyright ©1998-2020 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.modifier;

import com.trollworks.gcs.utility.Fixed6;
import com.trollworks.gcs.utility.text.Enums;
import com.trollworks.gcs.utility.units.WeightUnits;

public enum ModifierWeightValueType {
    ADDITION {
        @Override
        public String format(Fraction fraction, boolean localized) {
            String str = localized ? fraction.toLocalizedString() : fraction.toString();
            return fraction.mNumerator.greaterThanOrEqual(Fixed6.ZERO) ? "+" + str : str;
        }
    }, PERCENTAGE_ADDER {
        @Override
        public String format(Fraction fraction, boolean localized) {
            String str = (localized ? fraction.toLocalizedString() : fraction.toString()) + "%";
            return fraction.mNumerator.greaterThanOrEqual(Fixed6.ZERO) ? "+" + str : str;
        }
    }, PERCENTAGE_MULTIPLIER {
        @Override
        Fraction adjustFraction(Fraction fraction) {
            return fraction.mNumerator.lessThanOrEqual(Fixed6.ZERO) ? new Fraction(new Fixed6(100), Fixed6.ONE) : fraction;
        }

        @Override
        public String format(Fraction fraction, boolean localized) {
            fraction = adjustFraction(fraction);
            return "x" + (localized ? fraction.toLocalizedString() : fraction.toString()) + "%";
        }
    }, MULTIPLIER {
        @Override
        Fraction adjustFraction(Fraction fraction) {
            return fraction.mNumerator.lessThanOrEqual(Fixed6.ZERO) ? new Fraction(Fixed6.ONE, Fixed6.ONE) : fraction;
        }

        @Override
        public String format(Fraction fraction, boolean localized) {
            fraction = adjustFraction(fraction);
            return "x" + (localized ? fraction.toLocalizedString() : fraction.toString());
        }
    };

    Fraction adjustFraction(Fraction fraction) {
        return fraction;
    }

    public abstract String format(Fraction fraction, boolean localized);

    public Fraction extractFraction(String in, boolean localized) {
        return adjustFraction(new Fraction(in, localized));
    }

    public static WeightUnits extractUnits(String in, WeightUnits def) {
        in = in.trim();
        for (WeightUnits units : WeightUnits.values()) {
            String text = Enums.toId(units);
            if (in.endsWith(text)) {
                return units;
            }
        }
        return def;
    }

    public static ModifierWeightValueType determineType(String in) {
        in = in.trim().toLowerCase();
        if (in.endsWith("%")) {
            if (in.startsWith("x")) {
                return PERCENTAGE_MULTIPLIER;
            }
            return PERCENTAGE_ADDER;
        }
        if (in.startsWith("x") || in.endsWith("x")) {
            return MULTIPLIER;
        }
        return ADDITION;
    }
}

File: model/gurps/equipment/modifier_weight_value_type.go
/*
 * Copyright ©1998-2022 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package equipment

import (
	"strings"

	"github.com/richardwilkes/gcs/v5/model/fxp"
)

// Format returns a formatted version of the value.
func (enum ModifierWeightValueType) Format(fraction fxp.Fraction) string {
	switch enum {
	case WeightAddition:
		return fraction.StringWithSign()
	case WeightPercentageAdder:
		return fraction.StringWithSign() + enum.String()
	case WeightPercentageMultiplier:
		if fraction.Numerator <= 0 {
			fraction.Numerator = fxp.Hundred
			fraction.Denominator = fxp.One
		}
		return WeightMultiplier.String() + fraction.String() + WeightPercentageAdder.String()
	case WeightMultiplier:
		if fraction.Numerator <= 0 {
			fraction.Numerator = fxp.One
			fraction.Denominator = fxp.One
		}
		return enum.String() + fraction.String()
	default:
		return WeightAddition.Format(fraction)
	}
}

// ExtractFraction from the string.
func (enum ModifierWeightValueType) ExtractFraction(s string) fxp.Fraction {
	s = strings.TrimLeft(strings.TrimSpace(s), WeightMultiplier.Key())
	for len(s) > 0 && (s[len(s)-1] < '0' || s[len(s)-1] > '9') {
		s = s[:len(s)-1]
	}
	fraction := fxp.NewFraction(s)
	revised := enum.EnsureValid()
	switch revised {
	case WeightPercentageMultiplier:
		if fraction.Numerator <= 0 {
			fraction.Numerator = fxp.Hundred
			fraction.Denominator = fxp.One
		}
	case WeightMultiplier:
		if fraction.Numerator <= 0 {
			fraction.Numerator = fxp.One
			fraction.Denominator = fxp.One
		}
	default:
	}
	return fraction
}

// DetermineModifierWeightValueTypeFromString examines a string to determine what type it is.
func DetermineModifierWeightValueTypeFromString(s string) ModifierWeightValueType {
	s = strings.ToLower(strings.TrimSpace(s))
	switch {
	case strings.HasSuffix(s, "%"):
		if strings.HasPrefix(s, "x") {
			return WeightPercentageMultiplier
		}
		return WeightPercentageAdder
	case strings.HasPrefix(s, "x") || strings.HasSuffix(s, "x"):
		return WeightMultiplier
	default:
		return WeightAddition
	}
}
